Sources/SwiftProtobuf/ProtobufTextEncoding.swift

```diff
@@ -0,0 +1,291 @@
+            encoder.endField()
+        }
+    }
+    
+    mutating public func visitMapField<KeyType: ProtobufMapKeyType, ValueType: ProtobufMapValueType>(fieldType: ProtobufMap<KeyType, ValueType>.Type, value: ProtobufMap<KeyType, ValueType>.BaseType, protoFieldNumber: Int, protoFieldName: String, jsonFieldName: String, swiftFieldName: String) throws  where KeyType.BaseType: Hashable {
+        encoder.startField(name: protoFieldName, tabLevel: tabLevel)
+        var arraySeparator = ""
+        encoder.append(text: "{")
+        for (k,v) in value {
+            encoder.append(text: arraySeparator)
+            KeyType.serializeTextMapKeyValue(encoder: &encoder, value: k)
+            encoder.append(text: ":")
+            try ValueType.serializeTextValue(encoder: &encoder, value: v)
+            arraySeparator = ","
+        }
+        encoder.append(text: "}")
+        encoder.endField()
+    }
+}
+
+
+public struct ProtobufTextEncoder {
+    fileprivate var text: [String] = []
+    public init() {}
+    public var result: String { return text.joined(separator: "") }
+    
+    mutating func append(text newText: String) {
+        text.append(newText)
+    }
+    mutating func appendTokens(tokens: [ProtobufTextToken]) {
+        for t in tokens {
+            switch t {
+            case .beginArray: append(text: "[")
+            case .beginObject: append(text: "{")
+            case .boolean(let v):
+                // Note that quoted boolean map keys get stored as .string()
+                putBoolValue(value: v, quote: false)
+            case .colon: append(text: ":")
+            case .comma: append(text: ",")
+            case .endArray: append(text: "]")
+            case .endObject: append(text: "}")
+            case .null: putNullValue()
+            case .number(let v): append(text: v)
+            case .string(let v): putStringValue(value: v)
+            }
+        }
+    }
+    mutating func startField(name: String, tabLevel: Int, dropColon:Bool = false) {
+        for _ in 0..<tabLevel {
+            append(text:"  ")
+        }
+        
+        if dropColon {
+            append(text: name + " ")
+        } else {
+            append(text: name + ": ")
+        }
+    }
+    mutating func endField() {
+        append(text: "\n")
+    }
+    public mutating func startObject() {
+        append(text: "{\n")
+    }
+    public mutating func endObject(tabLevel: Int) {
+        for _ in 0..<tabLevel {
+            append(text:"  ")
+        }
+
+        append(text: "}")
+    }
+    mutating func putNullValue() {
+        append(text: "null")
+    }
+    mutating func putFloatValue(value: Float, quote: Bool) {
+        putDoubleValue(value: Double(value), quote: quote)
+    }
+    mutating func putDoubleValue(value: Double, quote: Bool) {
+        if value.isNaN {
+            append(text: "\"NaN\"")
+        } else if !value.isFinite {
+            if value < 0 {
+                append(text: "\"-Infinity\"")
```

Comment:
The "-Infinity" and similar cases need to be tested against Google's C++ implementation to make sure this is the correct coding.